### ISSUE TITLE
Deployment error / warning related to CDI configuration and MonitoringCdiInterceptor

### ISSUE THREAD

**Issue by jncharpin** [javamelody/javamelody#634]
Hi,

I just upgraded my application to make use of java melody 1.67 and noticed that deployment fails under weblogic 12.2.1 during CDI deployment with the following stack trace.

This is an ear file embedding java melody in APP-INF/lib. 

Not sure whether I missed something in the documentation regarding configuration or dependencies.

```
weblogic.management.DeploymentException: CDI deployment failure:org/hibernate/jdbc/BatchingBatcher
	at com.oracle.injection.integration.CDIAppDeploymentExtension.initCdi(CDIAppDeploymentExtension.java:95)
	at com.oracle.injection.integration.CDIAppDeploymentExtension.activate(CDIAppDeploymentExtension.java:43)
	at weblogic.application.internal.flow.AppDeploymentExtensionFlow.activate(AppDeploymentExtensionFlow.java:39)
	at weblogic.application.internal.BaseDeployment$2.next(BaseDeployment.java:750)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.BaseDeployment.activate(BaseDeployment.java:260)
	at weblogic.application.internal.EarDeployment.activate(EarDeployment.java:66)
	at weblogic.application.internal.DeploymentStateChecker.activate(DeploymentStateChecker.java:165)
	at weblogic.deploy.internal.targetserver.AppContainerInvoker.activate(AppContainerInvoker.java:90)
	at weblogic.deploy.internal.targetserver.BasicDeployment.activate(BasicDeployment.java:265)
	at weblogic.deploy.internal.targetserver.BasicDeployment.activateFromServerLifecycle(BasicDeployment.java:481)
	at weblogic.management.deploy.internal.DeploymentAdapter$1.doActivate(DeploymentAdapter.java:53)
	at weblogic.management.deploy.internal.DeploymentAdapter.activate(DeploymentAdapter.java:202)
	at weblogic.management.deploy.internal.AppTransition$2.transitionApp(AppTransition.java:52)
	at weblogic.management.deploy.internal.ConfiguredDeployments$2.doItem(ConfiguredDeployments.java:684)
	at weblogic.management.deploy.internal.parallel.BucketInvoker.invoke(BucketInvoker.java:138)
	at weblogic.management.deploy.internal.ConfiguredDeployments.transitionAppsParallel(ConfiguredDeployments.java:692)
	at weblogic.management.deploy.internal.ConfiguredDeployments.activate(ConfiguredDeployments.java:341)
	at weblogic.management.deploy.internal.ConfiguredDeployments.deploy(ConfiguredDeployments.java:203)
	at weblogic.management.deploy.internal.DeploymentServerService.resume(DeploymentServerService.java:207)
	at weblogic.management.deploy.internal.DeploymentServerService.start(DeploymentServerService.java:129)
	at weblogic.server.AbstractServerService.postConstruct(AbstractServerService.java:76)
	at sun.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.hk2.utilities.reflection.ReflectionHelper.invoke(ReflectionHelper.java:1262)
	at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:332)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:374)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:471)
	at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:232)
	at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:85)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2020)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:114)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:693)
	at org.jvnet.hk2.internal.ThreeThirtyResolver.resolve(ThreeThirtyResolver.java:78)
	at org.jvnet.hk2.internal.ClazzCreator.resolve(ClazzCreator.java:211)
	at org.jvnet.hk2.internal.ClazzCreator.resolveAllDependencies(ClazzCreator.java:234)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:357)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:471)
	at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:232)
	at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:85)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2020)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:114)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:88)
	at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.oneJob(CurrentTaskFuture.java:1213)
	at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.run(CurrentTaskFuture.java:1144)
	at weblogic.work.SelfTuningWorkManagerImpl$WorkAdapterImpl.run(SelfTuningWorkManagerImpl.java:666)
	at weblogic.invocation.ComponentInvocationContextManager._runAs(ComponentInvocationContextManager.java:348)
	at weblogic.invocation.ComponentInvocationContextManager.runAs(ComponentInvocationContextManager.java:333)
	at weblogic.work.LivePartitionUtility.doRunWorkUnderContext(LivePartitionUtility.java:54)
	at weblogic.work.PartitionUtility.runWorkUnderContext(PartitionUtility.java:41)
	at weblogic.work.SelfTuningWorkManagerImpl.runWorkUnderContext(SelfTuningWorkManagerImpl.java:640)
	at weblogic.work.ExecuteThread.execute(ExecuteThread.java:406)
	at weblogic.work.ExecuteThread.run(ExecuteThread.java:346)
Caused By: java.lang.ClassNotFoundException: org.hibernate.jdbc.BatchingBatcher
	at weblogic.utils.classloaders.GenericClassLoader.findLocalClass(GenericClassLoader.java:1026)
	at weblogic.utils.classloaders.GenericClassLoader.findClass(GenericClassLoader.java:987)
	at weblogic.utils.classloaders.GenericClassLoader.doFindClass(GenericClassLoader.java:608)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:540)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:493)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:470)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at weblogic.utils.classloaders.GenericClassLoader.defineClassInternal(GenericClassLoader.java:1110)
	at weblogic.utils.classloaders.GenericClassLoader.defineClass(GenericClassLoader.java:1043)
	at weblogic.utils.classloaders.GenericClassLoader.findLocalClass(GenericClassLoader.java:1035)
	at weblogic.utils.classloaders.GenericClassLoader.findClass(GenericClassLoader.java:987)
	at weblogic.utils.classloaders.GenericClassLoader.doFindClass(GenericClassLoader.java:608)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:540)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:493)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:470)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.oracle.injection.provider.weld.BasicResourceLoader.classForName(BasicResourceLoader.java:40)
	at org.jboss.weld.bootstrap.AnnotatedTypeLoader.loadClass(AnnotatedTypeLoader.java:65)
	at org.jboss.weld.bootstrap.AnnotatedTypeLoader.loadAnnotatedType(AnnotatedTypeLoader.java:60)
	at org.jboss.weld.bootstrap.BeanDeployer.addClass(BeanDeployer.java:97)
	at org.jboss.weld.bootstrap.BeanDeployer.addClasses(BeanDeployer.java:135)
	at org.jboss.weld.bootstrap.BeanDeployment.createClasses(BeanDeployment.java:203)
	at org.jboss.weld.bootstrap.WeldStartup.startInitialization(WeldStartup.java:375)
	at org.jboss.weld.bootstrap.WeldBootstrap.startInitialization(WeldBootstrap.java:76)
	at com.oracle.injection.provider.weld.WeldInjectionContainer.initialize(WeldInjectionContainer.java:94)
	at com.oracle.injection.integration.CDIAppDeploymentExtension.initCdi(CDIAppDeploymentExtension.java:80)
	at com.oracle.injection.integration.CDIAppDeploymentExtension.activate(CDIAppDeploymentExtension.java:43)
	at weblogic.application.internal.flow.AppDeploymentExtensionFlow.activate(AppDeploymentExtensionFlow.java:39)
	at weblogic.application.internal.BaseDeployment$2.next(BaseDeployment.java:750)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.BaseDeployment.activate(BaseDeployment.java:260)
	at weblogic.application.internal.EarDeployment.activate(EarDeployment.java:66)
	at weblogic.application.internal.DeploymentStateChecker.activate(DeploymentStateChecker.java:165)
	at weblogic.deploy.internal.targetserver.AppContainerInvoker.activate(AppContainerInvoker.java:90)
	at weblogic.deploy.internal.targetserver.BasicDeployment.activate(BasicDeployment.java:265)
	at weblogic.deploy.internal.targetserver.BasicDeployment.activateFromServerLifecycle(BasicDeployment.java:481)
	at weblogic.management.deploy.internal.DeploymentAdapter$1.doActivate(DeploymentAdapter.java:53)
	at weblogic.management.deploy.internal.DeploymentAdapter.activate(DeploymentAdapter.java:202)
	at weblogic.management.deploy.internal.AppTransition$2.transitionApp(AppTransition.java:52)
	at weblogic.management.deploy.internal.ConfiguredDeployments$2.doItem(ConfiguredDeployments.java:684)
	at weblogic.management.deploy.internal.parallel.BucketInvoker.invoke(BucketInvoker.java:138)
	at weblogic.management.deploy.internal.ConfiguredDeployments.transitionAppsParallel(ConfiguredDeployments.java:692)
	at weblogic.management.deploy.internal.ConfiguredDeployments.activate(ConfiguredDeployments.java:341)
	at weblogic.management.deploy.internal.ConfiguredDeployments.deploy(ConfiguredDeployments.java:203)
	at weblogic.management.deploy.internal.DeploymentServerService.resume(DeploymentServerService.java:207)
	at weblogic.management.deploy.internal.DeploymentServerService.start(DeploymentServerService.java:129)
	at weblogic.server.AbstractServerService.postConstruct(AbstractServerService.java:76)
	at sun.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.hk2.utilities.reflection.ReflectionHelper.invoke(ReflectionHelper.java:1262)
	at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:332)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:374)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:471)
	at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:232)
	at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:85)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2020)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:114)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:693)
	at org.jvnet.hk2.internal.ThreeThirtyResolver.resolve(ThreeThirtyResolver.java:78)
	at org.jvnet.hk2.internal.ClazzCreator.resolve(ClazzCreator.java:211)
	at org.jvnet.hk2.internal.ClazzCreator.resolveAllDependencies(ClazzCreator.java:234)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:357)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:471)
	at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:232)
	at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:85)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2020)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:114)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:88)
	at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.oneJob(CurrentTaskFuture.java:1213)
	at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.run(CurrentTaskFuture.java:1144)
	at weblogic.work.SelfTuningWorkManagerImpl$WorkAdapterImpl.run(SelfTuningWorkManagerImpl.java:666)
	at weblogic.invocation.ComponentInvocationContextManager._runAs(ComponentInvocationContextManager.java:348)
	at weblogic.invocation.ComponentInvocationContextManager.runAs(ComponentInvocationContextManager.java:333)
	at weblogic.work.LivePartitionUtility.doRunWorkUnderContext(LivePartitionUtility.java:54)
	at weblogic.work.PartitionUtility.runWorkUnderContext(PartitionUtility.java:41)
	at weblogic.work.SelfTuningWorkManagerImpl.runWorkUnderContext(SelfTuningWorkManagerImpl.java:640)
	at weblogic.work.ExecuteThread.execute(ExecuteThread.java:406)
	at weblogic.work.ExecuteThread.run(ExecuteThread.java:346)
```

I also noticed same type of warnings in JBoss 7EAP related to class not found (hibernate, spring, aop...), but not failing application deployment.

```
2017-05-23 11:29:23,619 WARN  [org.jboss.modules] (Weld Thread Pool -- 5) Failed to define class net.bull.javamelody.HibernateBatcherFactory in Module "deployment.xxx.ear:main" from Service Module Loader: java.lang.NoClassDefFoundError: Failed to link net/bull/javamelody/HibernateBatcherFactory (Module "deployment.xxx.ear:main" from Service Module Loader): org/hibernate/jdbc/BatcherFactory
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.jboss.modules.ModuleClassLoader.defineClass(ModuleClassLoader.java:446)
	at org.jboss.modules.ModuleClassLoader.loadClassLocal(ModuleClassLoader.java:274)
	at org.jboss.modules.ModuleClassLoader$1.loadClassLocal(ModuleClassLoader.java:78)
	at org.jboss.modules.Module.loadModuleClass(Module.java:605)
	at org.jboss.modules.ModuleClassLoader.findClass(ModuleClassLoader.java:190)
	at org.jboss.modules.ConcurrentClassLoader.performLoadClassUnchecked(ConcurrentClassLoader.java:363)
	at org.jboss.modules.ConcurrentClassLoader.performLoadClass(ConcurrentClassLoader.java:351)
	at org.jboss.modules.ConcurrentClassLoader.loadClass(ConcurrentClassLoader.java:93)
	at org.jboss.as.weld.WeldModuleResourceLoader.classForName(WeldModuleResourceLoader.java:68)
	at org.jboss.weld.bootstrap.AnnotatedTypeLoader.loadClass(AnnotatedTypeLoader.java:65)
	at org.jboss.weld.bootstrap.AnnotatedTypeLoader.loadAnnotatedType(AnnotatedTypeLoader.java:60)
	at org.jboss.weld.bootstrap.FastAnnotatedTypeLoader.loadAnnotatedType(FastAnnotatedTypeLoader.java:96)
	at org.jboss.weld.bootstrap.BeanDeployer.addClass(BeanDeployer.java:97)
	at org.jboss.weld.bootstrap.ConcurrentBeanDeployer$1.doWork(ConcurrentBeanDeployer.java:65)
	at org.jboss.weld.bootstrap.ConcurrentBeanDeployer$1.doWork(ConcurrentBeanDeployer.java:62)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:63)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:56)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at org.jboss.threads.JBossThread.run(JBossThread.java:320)

```

Thanks and regards,
Jean-Noel

**Comment by jncharpin** [comment]
Seems like this fix my issue **bean-discovery-mode="annotated"**

But I cannot guarantee that behavior is still the one expected as I'm not making use of this CDI interceptor.

`<?xml version="1.0" encoding="UTF-8"?>
<beans xmlns="http://java.sun.com/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/beans_1_1.xsd"
	bean-discovery-mode="annotated">
	<interceptors>
		<class>net.bull.javamelody.MonitoringCdiInterceptor</class>
	</interceptors>
</beans>
`

EJB monitoring using `MonitoringInterceptor` extension is still behaving as expected. 

Jean-Noel

**Comment by evernat** [comment]
Thanks for the issue, it is fixed now in master (03fead1) with bean-discovery-mode.
You can download a new build including the fix at:
https://javamelody.ci.cloudbees.com/job/javamelody/net.bull.javamelody$javamelody-core/
